manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,451 @@
+		deallocate: make(map[port]struct{}, len(endpoint.Ports)),
+	}
+
+	// now, deallocate everything in the old object
+	for _, p := range endpoint.Ports {
+		// skip publish mode ports
+		if p.PublishMode == api.PublishModeHost {
+			continue
+		}
+
+		prop.deallocate[port{p.Protocol, p.PublishedPort}] = struct{}{}
+	}
+
+	// and then allocate everything in the new object
+
+	// we'll do this in two steps. in the first step, allocate any ports that
+	// have a published port assigned already. this first step prevents us from
+	// choosing a published port for one port that another, later port wants to
+	// use
+	for _, p := range finalPorts {
+		// Skip all Host ports, which we take no action on
+		if p.PublishedPort == 0 || p.PublishMode == api.PublishModeHost {
+			continue
+		}
+		portObj := port{p.Protocol, p.PublishedPort}
+		if _, ok := pa.ports[portObj]; ok {
+			// check if we're deallocating this port
+			if _, ok := prop.deallocate[portObj]; !ok {
+				return nil, ErrPortInUse{portObj}
+			}
+		}
+		// also check that we haven't already tried to allocate this port for
+		// this particular endpoint
+		if _, ok := prop.allocate[portObj]; ok {
+			return nil, ErrPortInUse{portObj}
+		}
+
+		// now, mark this port as "in use" in the newPorts map.
+		prop.allocate[portObj] = struct{}{}
+	}
+
+	// now, this second go around, we'll choose all new publish ports to
+	// dynamically allocate for any port that doesn't have a publish port yet
+ports:
+	for _, p := range finalPorts {
+		// Again, skip all host ports
+		if p.PublishedPort != 0 || p.PublishMode == api.PublishModeHost {
+			continue
+		}
+		portObj := port{p.Protocol, DynamicPortStart}
+		// loop through the whole range of dynamic ports and select the first
+		// one available
+		for i := DynamicPortStart; i <= DynamicPortEnd; i++ {
+			portObj.port = i
+			if _, ok := pa.ports[portObj]; !ok {
+				// also check if the port has been assigned to some other
+				// port on this service
+				if _, ok := prop.allocate[portObj]; !ok {
+					// if the port is not in use, mark it in use, it's our's now,
+					// and continue to the next port
+					p.PublishedPort = i
+					prop.allocate[portObj] = struct{}{}
+					continue ports
+				}
+			}
+		}
+		// we're out of dynamic ports. check to see if we've deallocated any
+		// dynamic ports in this object
+		for deallocated := range prop.deallocate {
+			// is the protocol the same? is the deallocated port in the dynamic
+			// port range?
+			if deallocated.protocol == portObj.protocol &&
+				DynamicPortStart <= deallocated.port &&
+				deallocated.port <= DynamicPortEnd {
+				// are not we already reallocating this port?
+				if _, ok := prop.allocate[deallocated]; !ok {
+					// if all of the above, we can use that published port for
+					// this new port and move to the next port
+					portObj.port = deallocated.port
+					p.PublishedPort = deallocated.port
+					prop.allocate[portObj] = struct{}{}
+					continue ports
```

Comment:
What about removing it from `deallocate` instead of inserting it into `allocate`?
This would remove the need to check if it is in the allocate map, reduce the iterations in this loop to get to a solution and also simplify the commit by removing one add and one remove task.